tests/pytorch/kernel/test_flatten_kv_cache.py

```diff
@@ -0,0 +1,182 @@
+        yield max(num_blocks_per_input)
+
+    @pytest.fixture
+    def out_size(self, kv_lens):
+        yield sum(kv_lens)
+
+    @pytest.fixture
+    def kv_seqlens(self, kv_lens):
+        yield torch.tensor(kv_lens).cuda()
+
+    @pytest.fixture
+    def k_caches(self, batch_size, max_num_blocks, block_size, num_heads,
+                 head_dim, out_dtype):
+        shape = (batch_size * max_num_blocks, block_size, num_heads, head_dim)
+        yield torch.rand(shape, dtype=out_dtype, device='cuda')
+
+    @pytest.fixture
+    def v_caches(self, k_caches):
+        yield torch.rand_like(k_caches)
+
+    @pytest.fixture
+    def block_offsets(self, num_blocks_per_input):
+        batch_size = len(num_blocks_per_input)
+        max_num_blocks = max(num_blocks_per_input)
+        batch_ids = torch.arange(batch_size)
+        ret = torch.arange(max_num_blocks)
+        ret = batch_ids[:, None] + ret[None, :] * batch_size
+        yield ret.cuda()
+
+    @pytest.fixture
+    def gt(self, k_caches, v_caches, kv_lens, block_offsets, block_size,
+           num_heads, out_size, head_dim):
+        k_states = k_caches.new_empty(num_heads, out_size, head_dim)
+        v_states = v_caches.new_empty(num_heads, out_size, head_dim)
+        start_loc = 0
+        for kv_len, block_offs in zip(kv_lens, block_offsets):
+            remain_len = kv_len
+            for idx, _ in enumerate(range(0, kv_len, block_size)):
+                b_off = block_offs[idx]
+                block_len = min(block_size, remain_len)
+                end_loc = start_loc + block_len
+                k_block = k_caches[b_off, :block_len]
+                v_block = v_caches[b_off, :block_len]
+                k_states[:, start_loc:end_loc] = k_block.transpose(0, 1)
+                v_states[:, start_loc:end_loc] = v_block.transpose(0, 1)
+                start_loc = end_loc
+                remain_len -= block_len
+
+        yield k_states, v_states
+
+    def test_flatten_kv_cache(self, k_caches, v_caches, kv_seqlens,
+                              block_offsets, out_size, gt):
+        from lmdeploy.pytorch.kernels.cuda.flatten_kv_cache import \
+            flatten_kv_cache
+
+        k_states, v_states = flatten_kv_cache(k_caches,
+                                              v_caches,
+                                              kv_seqlens,
+                                              block_offsets,
+                                              out_size=out_size)
+        torch.testing.assert_close(k_states, gt[0])
+        torch.testing.assert_close(v_states, gt[1])
+
+
+def precise_round(x: torch.Tensor):
+    return x.sign() * (x.abs() + 0.5).floor()
+
+
+def quant(kv: torch.Tensor, nbits: int = 8):
+    """Quant kv on the head_dim."""
+    amax = kv.amax(dim=-1, keepdim=True)
+    amin = kv.amin(dim=-1, keepdim=True)
+    scales = (amax - amin) / (2**nbits - 1)
+    zeros = -amin / scales
+    q_kv = precise_round((kv - amin) / scales).to(torch.uint8)
```

Comment:
This might be slight different from triton kernel quantization. May refer to https://github.com/InternLM/lmdeploy/pull/2644